extension/spring-test/src/main/kotlin/org/camunda/bpm/extension/process_test_coverage/spring_test/ProcessEngineCoverageTestExecutionListener.kt

```diff
@@ -0,0 +1,221 @@
+     */
+    private fun initializeListeners(testContext: TestContext) {
+        val processEngineConfiguration = getProcessEngine(testContext).processEngineConfiguration as ProcessEngineConfigurationImpl
+        val bpmnParseListeners = processEngineConfiguration.customPostBPMNParseListeners
+        for (parseListener in bpmnParseListeners) {
+            if (parseListener is ElementCoverageParseListener) {
+                parseListener.setCoverageState(coverageCollector)
+            }
+        }
+
+        // Compensation event handler
+        val compensationEventHandler = processEngineConfiguration.getEventHandler("compensate")
+        if (compensationEventHandler is CompensationEventCoverageHandler) {
+            compensationEventHandler.setCoverageState(coverageCollector)
+        } else {
+            logger.warn("CompensationEventCoverageHandler not registered with process engine configuration!"
+                    + " Compensation boundary events coverage will not be registered.")
+        }
+    }
+
+    private fun getProcessEngine(testContext: TestContext) = testContext.applicationContext.getBean(ProcessEngine::class.java)
+
+    /**
+     * Logs and asserts the test method coverage and creates a graphical report.
+     *
+     * @param testContext test context
+     */
+    private fun handleTestMethodCoverage(testContext: TestContext) {
+        val suite = coverageCollector.activeSuite
+        val run = suite.getRun(testContext.testMethod.name) ?: return
+        val coveragePercentage = run.calculateCoverage(coverageCollector.getModels())
+
+        // Log coverage percentage
+        logger.info("${run.name} test method coverage is $coveragePercentage")
+        logCoverageDetail(run)
+
+        processEngineCoverageProperties.testMethodCoverageConditions()[run.name]?.let {
+            assertCoverage(coveragePercentage, it)
+        }
+    }
+
+    /**
+     * Logs the string representation of the passed suite object.
+     */
+    private fun logCoverageDetail(suite: Suite) {
+        if (logger.isDebugEnabled && processEngineCoverageProperties.detailedCoverageLogging()) {
+            logger.debug(suite.toString())
+        }
+    }
+
+    /**
+     * Logs the string representation of the passed run object.
+     */
+    private fun logCoverageDetail(run: Run) {
+        if (logger.isDebugEnabled && processEngineCoverageProperties.detailedCoverageLogging()) {
+            logger.debug(run.toString())
+        }
+    }
+
+    private fun assertCoverage(coverage: Double, conditions: List<Condition<Double>>) {
+        conditions.forEach { Assertions.assertThat(coverage).satisfies(it) }
+    }
+
+    private fun addClassCoverageAtLeast(percentage: Double) {
+        percentage.checkPercentage()
+        classCoverageAssertionConditions.add(
+                Condition<Double>({ p -> p >= percentage }, "matches if the coverage ratio is at least $percentage")
+        )
+    }
+
+}
+
+fun Double.checkPercentage() =
```

Comment:
I would suggest to write this in a slightly more developer-friendly way. There is a `require` which throws a better exception type.

Reply:
You are absolutely correct. Changed this and moved logic to builder.